hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/diskbalancer/DiskBalancerService.java

```diff
@@ -0,0 +1,487 @@
+        .build();
+  }
+
+  @Override
+  public BackgroundTaskQueue getTasks() {
+    BackgroundTaskQueue queue = new BackgroundTaskQueue();
+
+    if (!shouldRun) {
+      return queue;
+    }
+    metrics.incrRunningLoopCount();
+
+    if (shouldDelay()) {
+      metrics.incrIdleLoopExceedsBandwidthCount();
+      return queue;
+    }
+
+    int availableTaskCount = parallelThread - inProgressTasks.size();
+    if (availableTaskCount <= 0) {
+      LOG.info("No available thread for disk balancer service. " +
+          "Current thread count is {}.", parallelThread);
+      return queue;
+    }
+
+    for (int i = 0; i < availableTaskCount; i++) {
+      Pair<HddsVolume, HddsVolume> pair =
+          DiskBalancerUtils.getVolumePair(volumeSet, threshold, deltaSizes);
+      if (pair == null) {
+        continue;
+      }
+      HddsVolume sourceVolume = pair.getLeft(), destVolume = pair.getRight();
+      Iterator<Container<?>> itr = ozoneContainer.getController()
+          .getContainers(sourceVolume);
+      while (itr.hasNext()) {
+        ContainerData containerData = itr.next().getContainerData();
+        if (!inProgressContainers.contains(
+            containerData.getContainerID()) && containerData.isClosed()) {
+          queue.add(new DiskBalancerTask(containerData, sourceVolume,
+              destVolume));
+          inProgressContainers.add(containerData.getContainerID());
+          deltaSizes.put(sourceVolume, deltaSizes.getOrDefault(sourceVolume, 0L)
+              - containerData.getBytesUsed());
+          deltaSizes.put(destVolume, deltaSizes.getOrDefault(destVolume, 0L)
+              + containerData.getBytesUsed());
+          break;
+        }
+      }
+    }
+    if (queue.isEmpty()) {
+      metrics.incrIdleLoopNoAvailableVolumePairCount();
+    }
+    return queue;
+  }
+
+  private boolean shouldDelay() {
+    // We should wait for next AvailableTime.
+    if (Time.monotonicNow() <= nextAvailableTime.get()) {
+      return true;
+    }
+    // Calculate the next AvailableTime based on bandwidth
+    long bytesBalanced = balancedBytesInLastWindow.getAndSet(0L);
+
+    final int megaByte = 1024 * 1024;
+
+    // converting disk bandwidth in byte/millisec
+    float bytesPerMillisec = bandwidthInMB * megaByte / 1000f;
+    nextAvailableTime.set(Time.monotonicNow() +
+        ((long) (bytesBalanced / bytesPerMillisec)));
+    return false;
+  }
+
+  private class DiskBalancerTask implements BackgroundTask {
+
+    private HddsVolume sourceVolume;
+    private HddsVolume destVolume;
+    private ContainerData containerData;
+    private long containerId;
+
+    DiskBalancerTask(ContainerData containerData,
+        HddsVolume sourceVolume, HddsVolume destVolume) {
+      this.containerData = containerData;
+      this.sourceVolume = sourceVolume;
+      this.destVolume = destVolume;
+      this.containerId = containerData.getContainerID();
+    }
+
+    @Override
+    public BackgroundTaskResult call() {
```

Comment:
Can we simplify this function and probably move it to utils or make it part of KeyValueContainer.. class itself? I think we should just have an API call here.

Reply:
Could you illustrate more on this idea?

Since this operation is in a higher level of KeyValueContainer, and it needs the help of ContainerSet, I find it not simple to generalize this function to a util class.

Reply:
I would try looking for an appropriate class for this. I think the move logic should not be present in the disk balancer and there should be separate implementation for the different container versions.